### DISCUSSION TITLE
How to specify the "speed" parameter

### DISCUSSION THREAD

**Discussion by MimoKar** [bigcat88/pillow_heif#286] [Q&A] [answered]
### Describe the bug

Parameter speed is ignored during image encoding. Doesn't matter what value is set, it is using some default.

### Steps/Code to Reproduce

im.save(new_file,format='AVIF', quality=quality, alpha_quality=quality, speed=0)

### Expected Results

Speed of encoding will differ for different setting.

### Actual Results

Encoding speed is same.

### Versions

```shell
3.12.5 (tags/v3.12.5:ff3bc82, Aug  6 2024, 20:45:27) [MSC v.1940 64 bit (AMD64)]
Windows-11-10.0.27695-SP0
0.18.0
{'libheif': '1.18.1', 'HEIF': 'x265 HEVC encoder (3.5+1-f0c1022b6)', 'AVIF': 'AOMedia Project AV1 Encoder v3.9.1', 'encoders': {'x265': 'x265 HEVC encoder (3.5+1-f0c1022b6)', 'aom': 'AOMedia Project AV1 Encoder v3.9.1', 'mask': 'mask'}, 'decoders': {'libde265': 'libde265 HEVC decoder, version 1.0.15', 'aom': 'AOMedia Project AV1 Decoder v3.9.1'}}
```

**Comment by bigcat88** [comment]
Specifying `speed` parameter directly to `save` is not supported.

You can pass custom encoding parameters (i.e. `speed`) using `enc_params` kwarg during `save`

Here is a test for that:

https://github.com/bigcat88/pillow_heif/blob/dc33d11c8e38f1edbc438ba03230b58fd3e9f207/tests/write_test.py#L186-L193

You can check which parameters are supported by which encoder in the `libeheif` repository or in the encoder repository itself.

**Comment by bigcat88** [comment, marked as answer]
```shell
heif-enc -A --params     
Parameters for encoder `AOMedia Project AV1 Encoder 3.9.1`:
  realtime, default=false
  speed, default=6, [0;9]
  threads, default=12, [1;64]
  quality, default=50, [0;100]
  lossless, default=false
  chroma, default=420, { 420,422,444 }
  tune, default=ssim, { psnr,ssim }
  min-q, default=0, [0;63]
  max-q, default=63, [0;63]
  alpha-quality, [0;100]
  alpha-min-q, [0;63]
  alpha-max-q, [0;63]
  lossless-alpha, default=false
  ```
  
  
  ```python3
  
    Image.open("arrow.jpg").save("arrow1.avif", enc_params={"speed": 6})
    Image.open("arrow.jpg").save("arrow2.avif", enc_params={"speed": 1})
    Image.open("arrow.jpg").save("arrow3.avif", enc_params={"speed": 9})
```

**Comment by MimoKar** [comment]
alpha_quality is supported in save or via enc_params only?

And is auto tiling available/enabled ?

**Reply by bigcat88** [reply]
I am not sure about `alpha quality` parameter..

"tiling" currently is not supported, API for it was introduced only in latest `libheif`(`1.18`) that was published not so long ago.

If no one will make a PR to add "tiling support", I will do it myself at the end of autumn, most likely for the New Year.
